rust/lance-index/src/mem_wal.rs

```diff
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use crate::{Index, IndexType};
+use async_trait::async_trait;
+use lance_core::cache::DeepSizeOf;
+use lance_core::Error;
+use lance_table::format::pb;
+use lance_table::rowids::segment::U64Segment;
+use prost::Message;
+use roaring::RoaringBitmap;
+use serde::{Deserialize, Serialize};
+use snafu::location;
+use std::any::Any;
+use std::collections::{BTreeMap, HashMap};
+use std::sync::Arc;
+
+pub const MEM_WAL_INDEX_NAME: &str = "__lance_mem_wal";
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Serialize, Deserialize, DeepSizeOf)]
+pub struct MemWalId {
+    pub region: String,
+    pub generation: u64,
+}
+
+impl From<&MemWalId> for pb::mem_wal_index_details::MemWalId {
+    fn from(mem_wal: &MemWalId) -> Self {
+        Self {
+            region: mem_wal.region.clone(),
+            generation: mem_wal.generation,
+        }
+    }
+}
+
+impl TryFrom<pb::mem_wal_index_details::MemWalId> for MemWalId {
+    type Error = Error;
+
+    fn try_from(mem_wal: pb::mem_wal_index_details::MemWalId) -> lance_core::Result<Self> {
+        Ok(Self {
+            region: mem_wal.region.clone(),
+            generation: mem_wal.generation,
+        })
+    }
+}
+
+impl MemWalId {
+    pub fn new(region: &str, generation: u64) -> Self {
+        Self {
+            region: region.to_owned(),
+            generation,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Serialize, Deserialize, DeepSizeOf)]
+pub struct MemWal {
+    pub id: MemWalId,
+    pub mem_table_location: String,
+    pub wal_location: String,
+    pub wal_entries: Vec<u8>,
```

Comment:
Question: What is the motivation for storing an encoded `U64Segment`? It seems that every access would require decoding or encoding the segment again.

Reply:
U64Segment is missing a few important traits like Serializable, PartialOrd, Hash. Because when we use the `wal_entries` it's usually because we are writing a new one, so it is just a one time decoding and encoding anyway, so I did not bother optimizing that part yet.